RosettaJdbi3/src/main/java/com/hubspot/rosetta/jdbi3/BindWithRosetta.java

```diff
@@ -0,0 +1,58 @@
+package com.hubspot.rosetta.jdbi3;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jdbi.v3.sqlobject.customizer.SqlStatementCustomizerFactory;
+import org.jdbi.v3.sqlobject.customizer.SqlStatementCustomizingAnnotation;
+import org.jdbi.v3.sqlobject.customizer.SqlStatementParameterCustomizer;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.hubspot.rosetta.RosettaBinder;
+import com.hubspot.rosetta.jdbi3.BindWithRosetta.RosettaBinderFactory;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+@SqlStatementCustomizingAnnotation(RosettaBinderFactory.class)
+public @interface BindWithRosetta {
+  String prefix() default "";
+
+  class RosettaBinderFactory implements SqlStatementCustomizerFactory {
+
+    @Override
+    public SqlStatementParameterCustomizer createForParameter(
+        Annotation annotation,
+        Class<?> sqlObjectType,
+        Method method,
+        Parameter param,
+        int index,
+        Type paramType
+    ) {
+      return (stmt, arg) -> {
+        ObjectMapper objectMapper = RosettaObjectMapperOverride.resolve(stmt.getContext());
+
+        JsonNode node = objectMapper.valueToTree(arg);
+        String prefix = ((BindWithRosetta) annotation).prefix();
+
+        if (node.isValueNode() || node.isArray()) {
+          node = objectMapper.createObjectNode().set(prefix.isEmpty() ? "it" : prefix, node);
+          prefix = "";
+        }
+
+        Map<String, Object> namedValues = new HashMap<>();
+        RosettaBinder.INSTANCE.bind(prefix, node, namedValues::put);
```

Comment:
We've found that eagerly binding every attribute can to cause problems. e.g. if your object contains a property of some type for which there is no argument factory registered. This may cause a binding problem, even if the SQL statement never attempts to use that property.

For composite bindings, I would recommend implementing `NamedArgumentFinder` and calling `stmt.bindNamedArgumentFinder()`. Then you don't have to enumerate properties, but just provide the ones that the SQL statement asks for by name.

Reply:
This is just putting the values into a local map which is then bound with `SqlStatement#bindMap`, which it looks like uses a `NamedArgumentFinder`. Also, the object has been converted to a JSON tree using Jackson, so we know that these values must be one of `null` literal, boolean, number, string, or byte[] (all of which have built-in argument factories)
https://github.com/HubSpot/Rosetta/blob/a4cbbe6c387feaa45fae03dd89fd62a31b217040/RosettaCore/src/main/java/com/hubspot/rosetta/RosettaBinder.java#L39-L51